blogs/docker-ci-tricks-2/blog.md

```diff
@@ -0,0 +1,209 @@
+
+## Caching the images: make use of Docker tags
+
+It would be nice to cache the generated images, so that CI doesn't build the same stuff over and over again.
+By the way, the purpose of the extreme image size optimization outlined [in the previous part](https://medium.com/virtuslab/nifty-docker-tricks-for-your-ci-vol-1-c4a36d2192ea) is also to facilitate caching.
+
+Think for a moment what exactly causes one generated image to be different from another.
+We obviously have to take into account the versions of Git and Python &mdash; passing different combinations of these will surely result in a different final image.
+Therefore, the Docker image tag should include the information about these versions in order to make the caching possible.
+
+But with respect to the project files... since it's `ci/tox/build-context` that's passed as build context,
+the Dockerfile doesn't know anything about the files from outside that directory!
+This means that even though other files change as the project is being developed,
+we could theoretically use the same image if `ci/tox/build-context` remains untouched.
+Note that the build context usually changes very rarely compared to the actual source files.
+
+There's a catch here, though.
+The build context is not the only thing that can affect the final Docker image.
+A change to the `Dockerfile` itself can obviously lead to a different result,
+but modification to `docker-compose.yml` and even the scripts that run `docker-compose` may also influence the result e.g. by changing environment variables and/or values passed as arguments.
+But since all those files reside under `ci/tox/`, this doesn't make things much worse &mdash;
+the entire resulting build image depends only on the contents of `ci/tox/` directory instead of `ci/tox/build-context/`
+(with one exception that packages installed via `apt-get` can also get updated over time in their respective APT repositories).
+
+Therefore, the image can be uniquely identified by the hash of its `ci/tox/` directory.
+We don't even need to derive that hash manually, because git computes an SHA-1 hash for each object it stores.
+It's a well known fact that every commit has a unique hash, but actually SHA-1 hashes are also derived for each file (called a _blob_ in gitspeak) and each directory (called a _tree_).
+The hash of a tree is based on the hashes of all of its nested blobs and subtrees.
+More details can be found in [this slide deck on git internals (aka "git's guts")](https://slides.com/plipski/git-internals/).
+For our use case, it means that any change to the files inside `ci/tox/` will yield a new Docker image tag.
+
+To get the hash of a given directory within the current commit (HEAD),
+we can use a plumbing command called `rev-parse`:
+
+```shell script
+git rev-parse HEAD:ci/tox
+```
+
+The argument has the form `<revision>:<path>`, same as in `git show`.
+
+Note that this hash is not the same as the hash of the resulting Docker image.
```

Comment:
I'd **emphasize** the word "not" here, in the spirit of https://virtuslab.github.io/tips/#teamwork/communication-good-practices